### PR TITLE
Added default support

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Solves https://github.com/everit-org/json-schema/issues/91 and can be used only with a flag :smile:

### CODE REVIEW

**Comment by erosb** [comment]
Hello @slinkydeveloper ,

thank you for putting effort into this. At the first glance my most serious concern about this PR is that it handles default values only if it is a member of an object, and there are some cases when the root value is not an object. For example if your root schema is
```
{
  "type": "string",
  "default": "empty"
}
```
and the input json is `null`, then probably the library should signal back `"empty"` as a default value somehow to the caller side. But that raises yet another question - how to provide the default value back to the caller? The signature of the `void validate(Object)` method doesn't provide any way, but changing it to return some data holder would be a backward-compatibility breaking change.

On the other hand, we can also ignore this impedance, and say that with this patch we cover the vast majority of usecases of the "default" keyword, so it is good-enough.

I will think about it.

I'm going to leave some comments in code review too, soon.

File: core/src/main/java/org/everit/json/schema/loader/SchemaLoader.java [hunk 1]
```diff
@@ -178,6 +180,11 @@ SchemaLoaderBuilder pointerToCurrentObj(List<String> pointerToCurrentObj) {
             return this;
         }
 
+        SchemaLoaderBuilder useDefaults(boolean useDefaults) {
```
Comment (erosb) [inline core/src/main/java/org/everit/json/schema/loader/SchemaLoader.java]: Would you please add some javadoc? Also, this seems to be the toggle point of the feature, so it would be necessary to make the method `public`, otherwise users won't be able to use this feature. I know there are some builder methods here which are package-private, but unlike this one, those are only for internal use by the library.

File: core/src/main/java/org/everit/json/schema/loader/JsonValue.java [hunk 2]
```diff
@@ -234,4 +234,17 @@ public Integer requireInteger() {
         }
         throw ls.createSchemaException(typeOfValue(), Integer.class);
     }
+
+    public static Object deepToOrgJson(JsonValue v) {
```
Comment (erosb) [inline core/src/main/java/org/everit/json/schema/loader/JsonValue.java]: Please make this package-private. I can see how is this necessary for now, but it is a bit redundant with `JsonValue#value()` and `JsonValue#unwrap()` (and their overrides in `JsonObject` and `JsonArray`) so this should be cleaned up in the future (and if you make this `public` then it can't be deleted in any 1.x versions).

File: core/src/main/java/org/everit/json/schema/loader/ArraySchemaLoader.java [hunk 3]
```diff
@@ -40,6 +40,8 @@ public ArraySchemaLoader(LoadingState ls, SchemaLoader defaultLoader) {
         ls.schemaJson().maybe("minItems").map(JsonValue::requireInteger).ifPresent(builder::minItems);
         ls.schemaJson().maybe("maxItems").map(JsonValue::requireInteger).ifPresent(builder::maxItems);
         ls.schemaJson().maybe("uniqueItems").map(JsonValue::requireBoolean).ifPresent(builder::uniqueItems);
+        if (config.useDefaults)
+            ls.schemaJson().maybe("default").map(JsonValue::deepToOrgJson).ifPresent(builder::defaultValue);
```
Comment (erosb) [inline core/src/main/java/org/everit/json/schema/loader/ArraySchemaLoader.java]: See my notes in `#buildBooleanSchema()`

File: core/src/main/java/org/everit/json/schema/loader/SchemaLoader.java [hunk 4]
```diff
@@ -404,8 +423,13 @@ public SchemaLoader(SchemaLoaderBuilder builder) {
         }
     }
 
+    private BooleanSchema.Builder buildBooleanSchema() {
```
Comment (erosb) [inline core/src/main/java/org/everit/json/schema/loader/SchemaLoader.java]: I think now you can remove this method (just return `BooleanSchema.builder()` in line 414)

File: core/src/test/resources/org/everit/jsonvalidator/testschemas.json [hunk 5]
```diff
@@ -543,5 +543,44 @@
         }
       }
     }
+  },
+  "defaultsTest": {
```
Comment (erosb) [inline core/src/test/resources/org/everit/jsonvalidator/testschemas.json]: Would you please also add a test with `"default": null` ? The tricky thing there is that the org.json library represents null values with `JSONObject#NULL` instead of a null reference, so maybe you will have to improve the `#deepToOrgJson()` method to cover it.

File: core/src/main/java/org/everit/json/schema/loader/StringSchemaLoader.java [hunk 6]
```diff
@@ -18,25 +18,30 @@
...
         this.formatValidators = unmodifiableMap(requireNonNull(formatValidators, "formatValidators cannot be null"));
+        this.useDefault = useDefault;
     }
 
     public StringSchema.Builder load() {
         StringSchema.Builder builder = StringSchema.builder();
         ls.schemaJson().maybe("minLength").map(JsonValue::requireInteger).ifPresent(builder::minLength);
         ls.schemaJson().maybe("maxLength").map(JsonValue::requireInteger).ifPresent(builder::maxLength);
+        if (useDefault)
+            ls.schemaJson().maybe("default").map(JsonValue::requireString).ifPresent(builder::defaultValue);
```
Comment (erosb) [inline core/src/main/java/org/everit/json/schema/loader/StringSchemaLoader.java]: Now this can be removed (you unified the loading in `SchemaLoader#load()`)
